File: src/main/java/com/jacuum/map/GeneratedMap.java
package com.jacuum.map;

import com.jacuum.algo.Direction;

public final class GeneratedMap implements GameMap {

    private final String hash;
    private final SizePreset size;
    private final boolean[][] floor; // [y][x]
    private final int startX;
    private final int startY;
    private final int totalFloor;

    public GeneratedMap(String hash, SizePreset size, boolean[][] floor, int startX, int startY) {
        this.hash   = hash;
        this.size   = size;
        this.floor  = floor;
        this.startX = startX;
        this.startY = startY;
        int count = 0;
        for (boolean[] row : floor)
            for (boolean cell : row)
                if (cell) count++;
        this.totalFloor = count;
    }

    @Override public String hash()         { return hash; }
    @Override public SizePreset size()     { return size; }
    @Override public int width()           { return floor[0].length; }
    @Override public int height()          { return floor.length; }
    @Override public boolean isFloor(int x, int y) { return inBounds(x, y) && floor[y][x]; }
    @Override public int startX()          { return startX; }
    @Override public int startY()          { return startY; }
    @Override public int totalFloorTiles() { return totalFloor; }

    @Override
    public boolean hasWall(int x, int y, Direction direction) {
        int nx = x + direction.dx();
        int ny = y + direction.dy();
        return !inBounds(nx, ny) || !floor[ny][nx];
    }

    private boolean inBounds(int x, int y) {
        return x >= 0 && y >= 0 && x < width() && y < height();
    }
}

File: src/main/kotlin/com/jacuum/map/GeneratedMap.kt
package com.jacuum.map

import com.jacuum.algo.Direction

class GeneratedMap(
    private val hash: String,
    private val size: SizePreset,
    private val floor: Array<BooleanArray>,
    private val startX: Int,
    private val startY: Int
) : GameMap {
    private val totalFloor: Int = floor.sumOf { row -> row.count { it } }

    override fun hash(): String = hash
    override fun size(): SizePreset = size
    override fun width(): Int = floor[0].size
    override fun height(): Int = floor.size
    override fun isFloor(x: Int, y: Int): Boolean = inBounds(x, y) && floor[y][x]
    override fun startX(): Int = startX
    override fun startY(): Int = startY
    override fun totalFloorTiles(): Int = totalFloor

    override fun hasWall(x: Int, y: Int, direction: Direction): Boolean {
        val nx = x + direction.dx()
        val ny = y + direction.dy()
        return !inBounds(nx, ny) || !floor[ny][nx]
    }

    private fun inBounds(x: Int, y: Int) = x >= 0 && y >= 0 && x < width() && y < height()
}
